### PR TITLE
implements #21620: allowing to import multiple modules with shared names

### LINKED ISSUE #21620: Ideally, this line would work on it's own, but any reference to `m21496` would cause an ambiguous error, no?
Ideally, this line would work on it's own, but any reference to `m21496` would cause an ambiguous error, no?

_Originally posted by @Menduist in https://github.com/nim-lang/Nim/pull/21500#discussion_r1133308601_

### PR DESCRIPTION
**Pull request by jmgomez**
fixes https://github.com/nim-lang/Nim/issues/21620
ref https://github.com/nim-lang/Nim/pull/21500#discussion_r1133308601

### CODE REVIEW

**Comment by jmgomez** [comment]
@Araq @ringabout let me know if you want to see something implement differently or a particular test case implemented. Tried to adhere as much as possible to the compiler's style

**Comment by jmgomez** [comment]
There is an issue on Lin. Will look into it 

EDIT: The issue was that I assumed that duplicated imports doesnt introduce a new module into the context. Added an extra check to prevent the module collide with itself fixed it.

File: compiler/lookups.nim [hunk 1]
```diff
@@ -634,7 +635,19 @@ proc qualifiedLookUp*(c: PContext, n: PNode, flags: set[TLookupFlag]): PSym =
         if m == c.module:
           result = strTableGet(c.topLevelScope.symbols, ident).skipAlias(n, c.config)
         else:
-          result = someSym(c.graph, m, ident).skipAlias(n, c.config)
+          var collides : bool
```
Comment (Araq) [inline compiler/lookups.nim]:
```suggestion
          var collides: bool
```

File: compiler/lookups.nim [hunk 2]
```diff
@@ -634,7 +635,19 @@ proc qualifiedLookUp*(c: PContext, n: PNode, flags: set[TLookupFlag]): PSym =
...
+          var collides: bool
+          for i in 0..c.imports.high:
+            if c.imports[i].m.name == m.name and c.imports[i].m.id in c.importModuleMap:
+              let realModuleId = c.importModuleMap[c.imports[i].m.id]
+              if realModuleId != c.importModuleMap[m.id]:
+                collides = true
+                break
+
+          if collides:
+            var amb : bool
```
Comment (Araq) [inline compiler/lookups.nim]:
```suggestion
            var amb: bool
```

**Comment by jmgomez** [comment]
@Araq when you have the time let me know if it's fine as it is so I can work on updating the manual

File: compiler/lookups.nim [hunk 3]
```diff
@@ -634,7 +635,19 @@ proc qualifiedLookUp*(c: PContext, n: PNode, flags: set[TLookupFlag]): PSym =
         if m == c.module:
           result = strTableGet(c.topLevelScope.symbols, ident).skipAlias(n, c.config)
         else:
-          result = someSym(c.graph, m, ident).skipAlias(n, c.config)
+          var collides: bool
+          for i in 0..c.imports.high:
```
Comment (juancarlospaco) [inline compiler/lookups.nim]:
```suggestion
          for i in 0 .. c.imports.high:
```

Comment (Araq) [reply compiler/lookups.nim]: O(n) for `module.foobar` where `n` is the number of imported modules? That's not good enough.

Comment (jmgomez) [reply compiler/lookups.nim]: It only runs on qualifications though and over locally imported modules. Shouldnt it be fine? Couldnt find any other hook. Tried with the module graph but it's empty at that point. 
Caching the name/ids pars or converting the seq of imported modules into a table is not an option, right? Any suggestions?

File: compiler/importer.nim [hunk 4]
```diff
@@ -245,6 +245,10 @@ proc importModuleAs(c: PContext; n: PNode, realModule: PSym, importHidden: bool)
     result.options.incl optImportHidden
   c.unusedImports.add((result, n.info))
   c.importModuleMap[result.id] = realModule.id
+  var modules = c.importModuleLookup.getOrDefault(realModule.name.h, newSeq[int]())
```
Comment (Araq) [inline compiler/importer.nim]: Why do you use the hash of the name? Use the id instead.

Comment (jmgomez) [reply compiler/importer.nim]: My bad. I assumed they were different. Updated

File: compiler/importer.nim [hunk 5]
```diff
@@ -245,6 +245,10 @@ proc importModuleAs(c: PContext; n: PNode, realModule: PSym, importHidden: bool)
     result.options.incl optImportHidden
   c.unusedImports.add((result, n.info))
   c.importModuleMap[result.id] = realModule.id
+  var modules = c.importModuleLookup.getOrDefault(realModule.name.id)
+  if realModule.id notin modules:
+    modules.add realModule.id
+    c.importModuleLookup[realModule.name.id] = modules
```
Comment (Araq) [inline compiler/importer.nim]: Use `c.importModuleLookup.mgetOrPut(realModule.name.id, @[]).add realModule.id` instead.

**Pull request merged**